lib/c-api/src/wasm_c_api/module.rs

```diff
@@ -115,3 +115,364 @@ pub unsafe extern "C" fn wasm_module_serialize(
+                wasm_store_t* store = wasm_store_new(engine);
+
+                wasm_byte_vec_t wat;
+                wasm_byte_vec_new_from_string(&wat, "(module)");
+                wasm_byte_vec_t* wasm = wat2wasm(&wat);
+
+                wasm_module_t* module = wasm_module_new(store, wasm);
+
+                assert(module);
+
+                wasm_byte_vec_delete(wasm);
+                wasm_module_delete(module);
+                wasm_store_delete(store);
+                wasm_engine_delete(engine);
+            }
+        })
+        .success();
+    }
+
+    #[test]
+    fn test_module_exports() {
+        (assert_c! {
+            #include <string.h>
+            #include "tests/wasmer_wasm.h"
+
+            void assert_exporttype_name(const wasm_exporttype_t* exporttype, const char* expected) {
+                assert(strncmp(wasm_exporttype_name(exporttype)->data, expected, strlen(expected)) == 0);
+            }
+
+            int main() {
+                wasm_engine_t* engine = wasm_engine_new();
+                wasm_store_t* store = wasm_store_new(engine);
+
+                wasm_byte_vec_t wat;
+                wasm_byte_vec_new_from_string(
+                    &wat,
+                    "(module\n"
+                    "  (func (export \"function\") (param i32 i64))\n"
+                    "  (global (export \"global\") i32 (i32.const 7))\n"
+                    "  (table (export \"table\") 0 funcref)\n"
+                    "  (memory (export \"memory\") 1))"
+                );
+                wasm_byte_vec_t* wasm = wat2wasm(&wat);
+
+                wasm_module_t* module = wasm_module_new(store, wasm);
+
+                assert(module);
+
+                wasm_exporttype_vec_t export_types;
+                wasm_module_exports(module, &export_types);
+
+                assert(export_types.size == 4);
+
+                {
+                    wasm_exporttype_t* export_type = export_types.data[0];
+                    assert_exporttype_name(export_type, "function");
+
+                    const wasm_externtype_t* extern_type = wasm_exporttype_type(export_type);
+                    assert(wasm_externtype_kind(extern_type) == WASM_EXTERN_FUNC);
+
+                    wasm_functype_t* func_type = wasm_externtype_as_functype((wasm_externtype_t*) extern_type);
+
+                    const wasm_valtype_vec_t* func_params = wasm_functype_params(func_type);
+                    assert(func_params->size == 2);
+                    assert(wasm_valtype_kind(func_params->data[0]) == WASM_I32);
+                    assert(wasm_valtype_kind(func_params->data[1]) == WASM_I64);
+                    wasm_valtype_vec_delete((wasm_valtype_vec_t*) func_params);
+
+                    const wasm_valtype_vec_t* func_results = wasm_functype_results(func_type);
+                    assert(func_results->size == 0);
+                    wasm_valtype_vec_delete((wasm_valtype_vec_t*) func_results);
```

Comment:
> ### wasm_functype_results()
> const wasm_valtype_vec_t* wasm_functype_results(const wasm_functype_t*)
> Returns the list of results of this function type.
> The returned memory is owned by the wasm_functype_t argument, the caller should not deallocate it.
 -- https://docs.wasmtime.dev/c-api/wasm_8h.html#a5ca8dc78814cb58f10be35b92e615974

Notably this is const-incorrectness, you ordinarily shouldn't free a const pointer. It would indicate either a bug in our code or in wasm.h.

Reply:
I wanted to report that to the WG, because `wasm_functype_results` actually returns a `const *` here.

Reply:
OK, the `wasm.h` declares the functions as is:

```c
WASM_API_EXTERN const wasm_valtype_vec_t* wasm_functype_params(const wasm_functype_t*);
WASM_API_EXTERN const wasm_valtype_vec_t* wasm_functype_results(const wasm_functype_t*);
```

The output result of both functions has no `own`, so the ownership is still to `wasm_functype_t`. We need to fix our code.

Reply:
Those memory leaks are fixed!